odoo/http.py

```diff
@@ -830,6 +841,133 @@ def touch(self):
+        except geoip2.errors.AddressNotFoundError:
+            return self.empty_city
+
+    @lazy_property
+    def _country_record(self):
+        if '_city_record' in vars(self):
+            return self._city_record
+        try:
+            return root.geoip_country_db.country(self.ip)
+        except geoip2.errors.AddressNotFoundError:
+            return self.empty_country
+
+    def __getattr__(self, attr):
+        if hasattr(self.empty_country, attr):
+            return getattr(self._country_record, attr)
+        if hasattr(self.empty_city, attr):
+            return getattr(self._city_record, attr)
+        raise AttributeError(f"{self} has no attribute {attr!r}")
+
+    # Deprecated dict API
+
+    _deprecated_keys = ['country_code', 'country_name', 'city', 'latitude',
+                        'longitude', 'region', 'time_zone']
+
+    def _is_ip_geolocalized(self):
+        self_dict = vars(self)
+        return (
+            ('_city_record' in self_dict and self._city_record is not self.empty_city) or
+            ('_country_record' in self_dict and self._country_record is not self.empty_country)
+        )
+
+    def __getitem__(self, item):
+        def warn(old_dict_item, new_expression):
+            warnings.warn(
+                DEPRECATED_GEOIP_API_MESSAGE.format(
+                    old_dict_item=old_dict_item,
+                    new_expression=new_expression,
+                ),
+                DeprecationWarning,
+                stacklevel=3,
+            )
+
+        if item == 'country_name':
+            warn('country_name', '(request.geoip.country.name or request.geoip.continent.name)')
+            return self.country.name or self.continent.name
+
+        if item == 'country_code':
+            warn('country_code', '(request.geoip.country.iso_code or request.geoip.continent.code)')
+            return self.country.iso_code or self.continent.code
+
+        if item == 'city':
+            warn('city', 'request.geoip.city.name')
+            return self.city.name
+
+        if item == 'latitude':
+            warn('latitude', 'request.geoip.location.latitude')
+            return self.location.latitude
+
+        if item == 'longitude':
+            warn('longitude', 'request.geoip.location.longitude')
+            return self.location.longitude
+
+        if item == 'region':
+            warn('region', '(request.geoip.subdivisions[0].iso_code if request.geoip.subdivisions else None)')
+            return self.subdivisions[0].iso_code if self.subdivisions else None
+
+        if item == 'time_zone':
+            warn('time_zone', 'request.geoip.location.time_zone')
+            return self.location.time_zone
+
+        raise KeyError(item)
+
+    def __iter__(self):
+        warnings.warn("The dictionnary GeoIP API is deprecated.", DeprecationWarning)
+        if self._is_ip_geolocalized():
+            return iter(self._deprecated_keys)
+        else:
+            return iter([])
```

Comment:
Why should it not iterate over keys but with None values?

Reply:
I removed that part, it now raises NotImplementedError